w3sper.js/src/protocol-driver/mod.js

```diff
@@ -1035,27 +1035,113 @@ export const withdraw = async (info) =>
+ * @param malloc - memory allocating function
+ * @param memcpy - memory copying function
+ * @returns {Promise<void>}
+ */
+async function setPointers(ptr_obj, data_ptr_name, len_ptr_name, buffer, malloc, memcpy) {
+    ptr_obj[data_ptr_name] = await malloc(buffer.byteLength);
+    let len = buffer.byteLength;
+    await memcpy(ptr_obj[data_ptr_name], buffer);
+
+    const len_buf = new Uint8Array(4);
+    new DataView(len_buf.buffer).setUint32(0, len, true);
+    ptr_obj[len_ptr_name] = await malloc(4);
+    await memcpy(ptr_obj[len_ptr_name], len_buf);
+}
 
-  if (!buffer) {
-    return null;
-  }
+/**
+ * Serializes given payload to rkyv format
+ * @param payload - object containing payload, e.g. memo or contract call data
+ * @param malloc - memory allocating function
+ * @param memcpy - memory copying function
+ * @param create_tx_data - low level function performing the actual serialization
+ * @returns {Promise<Uint8Array<ArrayBuffer>|null>} Buffer containing serialized data
+ */
+async function serializePayload(payload = null, malloc, memcpy, create_tx_data) {
+    if (!payload) return null;
+    const ptr = Object.create(null);
+    let code;
+    let ret;
+
+    if ('memo' in payload && payload.memo) {
+        let buffer = strToArray(payload.memo);
+        if (!buffer) return null;
+        await setPointers(ptr, "memo", "memo_len", buffer, malloc, memcpy);
+        ret = await malloc(4);
+        ptr.dummy_contract_id = await malloc(32);
+        code = await create_tx_data(
+            null,
+            null,
+            null,
+            null,
+            ptr.dummy_contract_id,
+            ptr.memo_len,
+            ptr.memo,
+            ret
+        )
+    } else if ('fnName' in payload && 'fnArgs' in payload && 'contractId' in payload) {
+        // fn_name
+        let fn_name_arg = strToArray(payload.fnName);
+        if (!fn_name_arg) return null;
+        await setPointers(ptr, "fn_name", "fn_name_len", fn_name_arg, malloc, memcpy);
+        // fn_args
+        const fn_args_buffer = new Uint8Array(payload.fnArgs);
+        await setPointers(ptr, "fn_args", "fn_args_len", fn_args_buffer, malloc, memcpy);
+        // contract_id
+        const contract_id_buffer = new Uint8Array(payload.contract_id);
+        ptr.contract_id = await malloc(contract_id_buffer.byteLength);
+        await memcpy(ptr.contract_id, contract_id_buffer);
+        ret = await malloc(4);
+        code = await create_tx_data(
+            ptr.fn_name_len,
+            ptr.fn_name,
+            ptr.fn_args_len,
+            ptr.fn_args,
+            ptr.contract_id,
+            null,
+            null,
+            ret
+        )
+    } else return null;
+
+    if (code > 0) throw DriverError.from(code);
 
-  const memoBuffer = new Uint8Array(1 + buffer.byteLength);
-  memoBuffer[0] = 3; // Memo type
-  memoBuffer.set(buffer, 1);
+    let ret_ptr = new DataView((await memcpy(null, ret, 4)).buffer).getUint32(
```

Comment:
These three variables (`ret_ptr`, `ret_len`, `ret_buf`) should be declared with `const` as they are never reassigned.

Reply:
done